File: Frameworks/MetalPetal/MTIAlphaType.m
//
//  MTIAlphaType.m
//  Pods
//
//  Created by Yu Ao on 23/10/2017.
//

#import "MTIAlphaType.h"
#import "MTIImage.h"

NSString * MTIAlphaTypeGetDescription(MTIAlphaType alphaType) {
    switch (alphaType) {
        case MTIAlphaTypePremultiplied:
            return @"Premultiplied";
        case MTIAlphaTypeNonPremultiplied:
            return @"NonPremultiplied";
        case MTIAlphaTypeAlphaIsOne:
            return @"AlphaIsOne";
        default:
            return @"UnknownAlphaType";
    }
}

@interface MTIAlphaTypeHandlingRule ()

@property (nonatomic,copy,readonly) MTIAlphaTypeHandlingOutputAlphaTypeRule outputAlphaTypeHandler;
@property (nonatomic,readonly) MTIAlphaType outputAlphaType;

@end

@implementation MTIAlphaTypeHandlingRule

- (instancetype)initWithAcceptableAlphaTypes:(NSArray<NSNumber *> *)acceptableAlphaTypes outputAlphaTypeHandler:(MTIAlphaTypeHandlingOutputAlphaTypeRule)outputAlphaTypeHandler {
    if (self = [super init]) {
        _acceptableAlphaTypes = acceptableAlphaTypes;
        _outputAlphaTypeHandler = [outputAlphaTypeHandler copy];
    }
    return self;
}

- (instancetype)initWithAcceptableAlphaTypes:(NSArray<NSNumber *> *)acceptableAlphaTypes outputAlphaType:(MTIAlphaType)outputAlphaType {
    if (self = [super init]) {
        _acceptableAlphaTypes = acceptableAlphaTypes;
        _outputAlphaType = outputAlphaType;
    }
    return self;
}

- (id)copyWithZone:(NSZone *)zone {
    return self;
}

- (BOOL)canAcceptAlphaType:(MTIAlphaType)alphaType {
    return [self.acceptableAlphaTypes containsObject:@(alphaType)];
}

- (MTIAlphaType)outputAlphaTypeForInputAlphaTypes:(NSArray<NSNumber *> *)inputAlphaTypes {
    if (_outputAlphaTypeHandler) {
        return _outputAlphaTypeHandler(inputAlphaTypes);
    } else {
        return _outputAlphaType;
    }
}

- (MTIAlphaType)outputAlphaTypeForInputImages:(NSArray<MTIImage *> *)inputImages {
    if (_outputAlphaTypeHandler) {
        NSMutableArray *alphaTypes = [NSMutableArray arrayWithCapacity:inputImages.count];
        for (MTIImage *image in inputImages) {
            [alphaTypes addObject:@(image.alphaType)];
        }
        NSParameterAssert([[NSSet setWithArray:alphaTypes] isSubsetOfSet:[NSSet setWithArray:_acceptableAlphaTypes]]);
        return [self outputAlphaTypeForInputAlphaTypes:alphaTypes];
    } else {
        return _outputAlphaType;
    }
}

+ (MTIAlphaTypeHandlingRule *)generalAlphaTypeHandlingRule {
    static MTIAlphaTypeHandlingRule * generalAlphaTypeHandlingRule;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        generalAlphaTypeHandlingRule = [[MTIAlphaTypeHandlingRule alloc] initWithAcceptableAlphaTypes:@[@(MTIAlphaTypeNonPremultiplied),@(MTIAlphaTypeAlphaIsOne)] outputAlphaType:MTIAlphaTypeNonPremultiplied];
    });
    return generalAlphaTypeHandlingRule;
}

+ (MTIAlphaTypeHandlingRule *)passthroughAlphaTypeHandlingRule {
    static MTIAlphaTypeHandlingRule * passthroughAlphaTypeHandlingRule;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        passthroughAlphaTypeHandlingRule = [[MTIAlphaTypeHandlingRule alloc] initWithAcceptableAlphaTypes:@[@(MTIAlphaTypeNonPremultiplied),@(MTIAlphaTypeAlphaIsOne),@(MTIAlphaTypePremultiplied)] outputAlphaTypeHandler:^MTIAlphaType(NSArray<NSNumber *> * _Nonnull inputAlphaTypes) {
            NSAssert(inputAlphaTypes.count == 1, @"");
            return [inputAlphaTypes.firstObject integerValue];
        }];
    });
    return passthroughAlphaTypeHandlingRule;
}

@end

File: Frameworks/MetalPetal/MTIAlphaType.swift
//
//  MTIAlphaType.swift
//  MetalPetal
//
//  Created by Yu Ao on 23/10/2017.
//

import Foundation

extension MTIAlphaTypeHandlingRule {
    
    public var acceptableAlphaTypes: [MTIAlphaType] {
        return self.__acceptableAlphaTypes.compactMap({ value in
            return MTIAlphaType(rawValue: value.intValue)
        })
    }
    
    public convenience init(acceptableAlphaTypes: [MTIAlphaType], outputAlphaType: MTIAlphaType) {
        self.init(__acceptableAlphaTypes: acceptableAlphaTypes.map({ NSNumber(value: $0.rawValue) }), outputAlphaType: outputAlphaType)
    }
    
    public convenience init(acceptableAlphaTypes: [MTIAlphaType], _ handler: @escaping ([MTIAlphaType]) -> MTIAlphaType) {
        self.init(__acceptableAlphaTypes: acceptableAlphaTypes.map({ NSNumber(value: $0.rawValue) }), outputAlphaTypeHandler: { types in
            return handler(types.map({ MTIAlphaType.init(rawValue: $0.intValue)! }))
        })
    }
    
    public convenience init(_ handler: @escaping ([MTIAlphaType]) -> MTIAlphaType) {
        self.init(__acceptableAlphaTypes: [MTIAlphaType.premultiplied, MTIAlphaType.nonPremultiplied, MTIAlphaType.alphaIsOne].map({ NSNumber(value: $0.rawValue) }), outputAlphaTypeHandler: { types in
            return handler(types.map({ MTIAlphaType.init(rawValue: $0.intValue)! }))
        })
    }
    
    public func outputAlphaType(forInputAlphaTypes inputAlphaTypes: [MTIAlphaType]) -> MTIAlphaType {
        return self.__outputAlphaType(forInputAlphaTypes: inputAlphaTypes.map({ NSNumber(value: $0.rawValue) }))
    }
    
}

extension MTIAlphaType: CustomStringConvertible {
    
    public var description: String {
        return MTIAlphaTypeGetDescription(self)
    }
    
}
